### ISSUE TITLE
--deploy-url will have no effect on the HTML

### ISSUE THREAD

**Issue by yordis** [angular/angular-cli#6666]
### Bug Report or Feature Request (mark with an `x`)
```
- [ ] bug report -> please search issues before submitting
- [ x ] feature request
```

### Versions.

@angular/cli: 1.0.1
node: 8.0.0
os: darwin x64
@angular/animations: 4.1.3
@angular/common: 4.1.3
@angular/compiler: 4.1.3
@angular/core: 4.1.3
@angular/forms: 4.1.3
@angular/http: 4.1.3
@angular/material: 2.0.0-beta.6
@angular/platform-browser: 4.1.3
@angular/platform-browser-dynamic: 4.1.3
@angular/router: 4.1.3
@angular/cli: 1.0.1
@angular/compiler-cli: 4.1.3

### Repro steps.

Related to https://github.com/angular/angular-cli/issues/4517#issuecomment-286252628

### Desired functionality.

The CLI transform the HTML as well

### Mention any other details that might be useful.

I am using `S3` on AWS this is kind of required for me unless I used `CloudFront`, I think.

**Comment by vukasin0** [comment]
+1 for this

**Comment by yordis** [comment]
@Brocco could we do something like `i18n` for the URL on the HTML, because I can see the big issues on it but we could use the same strategy in the worse scenario.

**Comment by sgwatgit** [comment]
This would be great for the `/dist/index.html`. I edit this file before deploying because my app lives under `domain.com/site/sub/path/index.html` and the default `src="main.bundle.js"` paths won't work in this situation.

It's not a big deal to edit this file with each deployment but it would be nice to have it handled by the cli with the insertion of the `<base />` element.

**Comment by filipesilva** [comment]
This is something I'd like to have if we can find a safe way to do it. For reference, there's some more context in https://github.com/angular/angular-cli/issues/4517#issuecomment-326970581.

**Comment by rajdeep26** [comment]
Any update here?

**Comment by lqclester** [comment]
how to fix it?

**Comment by RSginer** [comment]
+1 fix it please!

i have an app with different base href and the sass is compiling correctly the routes but not in html link, src, etc

**Comment by leichenau** [comment]
it's not only inconvenient, the code won't work.

**Comment by timelyxyz** [comment]
I'm using ng5 now. My workaround is to add a prefix `{{baseHref}}` of each src in HTML template.
```HTML
<img src="{{baseHref}}/assets/sample.png">
```
@RSginer

**Comment by timbru31** [comment]
@timelyxyz this does not work for AOT :(

**Comment by christophedemey** [comment]
This is horrible.

**Comment by 100cm** [comment]
pipe this  plz， 我的手都快加累死了。

**Comment by davidmcdougs** [comment]
Best solution for now is to use the background-image css property. The cli will add the base href attribute to this.

**Comment by gazsiadam** [comment]
Workaround until Angular takes the effort to fix, is to add a property to your environment variable, like deployUrl: '/xyz/', then for all 'assets' objects, you can add the following logic in html
`<img src="{{getEnv().deployUrl}}assets/img/logo.svg" class="login-logo">`

**Comment by csutorasr** [comment]
I went through the code of the Ivy rendering engine, compilers and CLI.

The main problem is that the url(...) replacement in the css files are done by webpack. The Html template is parsed by Angular compiler.

I think an AssetUrlVisitor could solve this problem. It should add the route as string concat. That would make img[src] work attribute without any binding. If it has binding the url should be added before the interpolation runs.

The visitor should look for img with src and make the changes to it. Probably there are more scenarios for example src-set for the future. Extension with the visitor would be easy. (Ivy is really great!)

The new url config should have a new name, so it could be introduced without breaking changes. E. g. assetUrl. Then in later versions could be merged with deployUrl.

The policy how to change the links could be the same as the algorithm made by webpack, for the best consistency.

**Comment by anvlkv** [comment]
Hi,

I have a hybrid app. Not just angular hybrid, but a lot of stuff. So my deployUrl is virtual directory in IIS. I can not, or couldn't figure out how to, change baseHref, because for the page url '/' is just the base i need.

I have pretty much the same setup on dev machine as on server. So I could fix the assets paths thing by adding deployUrl to my src attributes, which is... yes on of those "fixes".

However now, I'm using storybook and it doesn't have to know about deployUrl, so all my assets are 404.

Is there way to make it work without touching baseHref?

**Comment by s4m0r4m4** [comment]
One quick note - if, after build, the assets are located in the base folder of index.html, you can use relative paths to access them by prefixing with ".":
`<img src="./assets/img/logo.svg">`

This seems to be working with the baseHref build flag, since the path is relative.

**Comment by fengjihua** [comment]
My workaround is to use nginx router rewrite.

nging.conf:

```
        location /assets/ {
            rewrite ^/assets/(.*)$ http://cdn.example.com/assets/$1 redirect;
        }
```

**Comment by rajat1saxena** [comment]
Has someone discovered any workaround to this? 

I don't want to append `baseHref` to all my images in HTML templates. I am playing around with custom webpack configuration, but nothing seems to be working as of now. 

Has anyone seen any progress with a custom Webpack configuration or anything for that matter?

**Comment by clydin** [comment]
A web browser will use the configured HTML base HREF value with all relative URLs.  As long as relative paths are used with a properly configured base HREF value, no custom configuration should be required.

The deploy URL option is entirely different and is best avoided where possible as it hard codes the path within the application (potentially in multiple locations).  An alternative for deploy URL that covers many situations is to use a combination of an HTML base HREF (with potentially not having one) and the Angular `APP_BASE_HREF` token ([docs](https://angular.io/api/common/APP_BASE_HREF)).  The later of which controls the base URL for the Angular router.

**Comment by rajat1saxena** [comment]
@clydin 
I want to serve images from a CDN and the rest of the app should point to `/`. Will your solution help me achieve that?

I tried setting `APP_BASE_HREF` to `cdn.example.com` and now my app automatically redirects to `localhost:4200/cdn.example.com`. Is this the expected behaviour?

**Comment by tonystarkdo** [comment]
This need to be fixed urgently. Don't you developer guys use CDN? Angular is really a outdated framework. @filipesilva

**Comment by xinthose** [comment]
In my case I had to change my src from 
`<mdb-card-img src="../../assets/404.png" alt="Card image cap"></mdb-card-img>`
to 
`<mdb-card-img src="assets/404.png" alt="Card image cap"></mdb-card-img>`
and `ng serve` properly used the path of `http://localhost:4200/eds4/assets/404.png`

**Comment by xiaody** [comment]
As a workaround, we change

```html
<!-- component.html -->
<img src="assets/arrow-left.png" />
```

to something like this

```ts
// component.ts
const arrowLeft = require('assets/arrow-left.png');

export class MyComponent implements OnInit {
  readonly arrowLeft = arrowLeft;
}
```
```html
<!-- component.html -->
<img [src]="arrowLeft" />
```

By requiring the image in ts, the file can be correctly loaded by file-loader, and so both file hash and deploy-url work. This pattern is also very common in React world.

**Comment by timbru31** [comment]
A potential workaround that worked for us, can be found here, too: https://stackoverflow.com/a/46493276/1902598
Basically set `APP_BASE_HREF` with the help of the `PlatformLocation.getBaseHrefFromDOM`

**Comment by neo** [comment]
I recently also ran into the same situation because of https://github.com/angular/angular-cli/issues/15295#issuecomment-520098529.
After trying a couple different ways, a solution I found that suits me well is using different `assets` in `angular.json` for `production`: https://angular.io/guide/workspace-config#asset-config (the "asset specification object" part)

**Comment by JonathanWylie** [comment]
@clydin is it possible to configure the APP_BASE_HREF from the angular.json file?
I would like to configure the two different baseHref and appBaseHref in the same place and from configuration.

**Comment by thescientist13** [comment]
Is there any way to decouple `APP_BASE_HREF` provider from `--base-href` somehow?

That way for the browser, it could respect the HTML tag for all relative static assets, lazy-loaded chunks, etc
```html
<base href="http://www.somecdn.com/"/>
```

But for Angular router, configuring the provider to respect a different "root", eg:
```javascript
providers: [
  {
      provide: APP_BASE_HREF, useValue: `/` // or even `${window.location.origin}/`
  }
]
```

---

So just curious to know from the maintainer's point of view, if there is a way forward through some sort of enhancement to Angular and / or the CLI (whether it's this idea or something completely different)

If so, I would be happy to take a look into implementing such a feature, opening an issue for discussion, and contributing.  Just want to make sure there is a path forward and would then be happy to put some time into working on it.

Thanks!

**Comment by vultix** [comment]
Here's my workaround:
```typescript
declare const __webpack_public_path__: string;
export function deployUrl(url: string): string {
  let publicPath = __webpack_public_path__;
  if (!publicPath.endsWith('/')) {
    publicPath += '/';
  }
  if (url.startsWith('/')) {
    url = url.slice(1);
  }

  return `${publicPath}${url}`;
}
```

In component:
```typescript
@Component({...})
export class MyComponent {
  deployUrl = deployurl;
}
```

In template:
```html
<img [src]="deployUrl('assets/image.png')"/>
```

**Comment by dgp1130** [comment]
As there is an awful lot of confusion around `--deploy-url` I think this is a good candidate to review and update documentation. Specifically this could include:
1. What use cases `--deploy-url` is intended for (and what it is not).
1. For several common use cases (such as CDN's), what configuration recipe (whether using `--deploy-url` or other means) makes the most sense and does what an application wants.

/cc @aikidave for visibility.

**Comment by andreElrico** [comment]
Copied from my [STACKOVERFLOW](https://stackoverflow.com/a/62431840/6852937) answer. Its a very annoying problem: Here is how I solve it:

When your app is for e.g. hosted under 
`www.domain.de/hehe`.

1. in `angular.json` add `"deployUrl": "./",`
2. in `src/index.html` inside `<head>` add/update `<base href="./">`
3. throughout your app you will use relative paths for assets and whatnot: `<img src="./assets/img/pictureOfMyCrazyWife.png">`

Hope this helps and saves some grey hair, couse it only looks good on George Clooney.

**Comment by CaselIT** [comment]
@andreElrico that does not seem to work if you are using routing
If you try to navigate from the base path, let's say `/base-path/` to `/base-path/router-path` then try to reload the page or copy the link to another tab, the application will fail to load since it tries to use  `/base-path/router-path` as base path

**Comment by andreElrico** [comment]
@CaselIT have not tried it with routing.

**Comment by mderazon** [comment]
The problem of not being to add cache-busting hash to images used in Angular templates is a serious issue many are dealing with.

I understand that this is out of scope for the CLI team at the moment, but I have seen some workarounds. I think it will be very helpful for the community if we had a "blessed" workaround for this from the cli team and get their view on this. I will summarize what i've seen so far.

If you want image cache-busting / fingerprinting you have these options:
1. Only load images from css using relative path - This works fine, but some cases you want an `<img>` tag in template.
2. [This comment ](https://github.com/angular/angular-cli/issues/6666#issuecomment-565357948) in this thread. Basically _require_ the image in your component through Webpack and inject into the template. There are other, perhaps more elegant solitions this one](https://stackoverflow.com/a/57319321/599912) but same concept. However, these were sort-of recommended against by @filipesilva in [this comment](https://github.com/angular/angular-cli/issues/17400#issuecomment-609744000).
3. Do some search-and-replace hack post build. I don't think anybody considers this option...
4. I've seen [a comment](https://github.com/angular/angular-cli/issues/3415#issuecomment-515157794) by @mgechev about using [ngx-build-plus](https://github.com/manfredsteyer/ngx-build-plus) for this purpose. Not sure how.

To summarize, I think it would be very helpful to get an answer from the cli team and perhaps add it to the cookbook/recipe/docs

**Comment by oliverdavidt** [comment]
@filipesilva I want to second @mderazon idea about having a "blessed" solution for this. It seems like the community is wasting a lot of time on something that is pretty common. I'd be willing to help with a solution, but I think we are looking to you guys for a little direction as to what you think should be done or info if you are already working on something.

**Comment by HDMuller** [comment]
I don't know if it is the same issue, because I use command --base-href of Angular.
But if it helps someone, I'm happy.

I used the "\" before the assets like this:
`<img [src]="'\assets/images/image.png'"/>`

And this worked perfectly for me.

**Comment by jorgeso** [comment]
This is really annoying. The `--base-href` parameter is available by injecting `APP_BASE_HREF`. I don't understand why `--deploy-url` can't be accessed the same way.

**Comment by BruneXX** [comment]
Also guys, someone tried this with angular-universal? If I use: 
`--deploy-url /myPath/` I'm receiving this erros in node...
`Unable to locate stylesheet: /var/www/html/dist/browser/**myPath**/styles.16d89a3cd69123f2ba9c.css`

Also the url set to the assets seems to be wrong..

`Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "text/html". Strict MIME type checking is enforced for module scripts per HTML spec.`

**Comment by Rishanthakumar** [comment]
According to https://github.com/angular/angular-cli/issues/6666#issuecomment-595974291 in this thread. I tried building the app using:

` ng build --c=qa --base-href https://your-assets-url/
`

and then in the app module to change the router base href:

`providers: [
  {
      provide: APP_BASE_HREF, useValue:  '${window.location.origin}/'  (replace single quotes )
  }
]`

This works fine for my scenario where I am serving the index.html in a different url and the assets from a cdn.
Correct me if I am doing anything wrong.

**Comment by maciejgunia** [comment]
> Also guys, someone tried this with angular-universal? If I use:
> `--deploy-url /myPath/` I'm receiving these erros in node...
> 
> `Unable to locate stylesheet: /var/www/html/dist/browser/**myPath**/styles.16d89a3cd69123f2ba9c.css`
> 
> Also the url set to the assets seems to be wrong..
> 
> `Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "text/html". Strict MIME type checking is enforced for module scripts per HTML spec.`

Did you manage to solve it? I have the same thing, it doesn't result in any app misbehaviour but it does bug me a bit.
